### PR TITLE
feat: middleware that parses the request body into a given struct

### LINKED ISSUE #8: Create a middleware to parse request body based on struct inputs
Currently, we can access the request body via `r.Body` in the controller and read out the data using a decoder `json.NewDecoder(r.Body).Decode(&struct{}{})` but this means we have to repeat this same action over and over in all controllers. 

We need a middleware that can use used in any route definition, take a struct, decode the body using the struct, and append the decoded data into the request context such that we can access it later on. 

By the word `used in any route definition`, I mean can be used with route functions of type `func (w http.ResponseWriter, r *http.Request)` as well as with route handlers of type `http.Handler`

Essentially, we will be needing two functions and they should be named `BodyH()` and `BodyF`

### PR DESCRIPTION
**Pull request by Funmi4194**

### CODE REVIEW

**Comment by opensaucerer** [comment]
@Funmi4194 This PR shows that the git-flow guide I shared was not correctly followed. Please close this MR and create another one that follows the git-flow guide. 

If you follow the git-flow guide, then the GitHub issue I created will also be linked to the PR automatically.

**Review by opensaucerer — CHANGES_REQUESTED** [review]
In this PR...there's a file called `bin/app` which shows that you probably made use of the `Makefile` to start the code... but you did not add `bin/` to the `.gitignore` file cause it to get added into your version history.... 

Please remove it and ignore it in git...we don't need to track the `app` or anything generated inside of `bin` as `bin` is, in this case, meant to contain only binary files generated by `go build` and we don't need that here...we only need it locally by `make run`

File: middleware/v1/body.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+
+	"github.com/opensaucerer/gotemplate/helper"
+	"github.com/opensaucerer/gotemplate/typing"
+)
+
+//BodyH is a middleware that parses the request body into the given struct
```
Comment (opensaucerer) [inline middleware/v1/body.go]: `BodyH` and `BodyF` both contain the same comment...how do I differentiate them? Let the comment illustrate their difference

```go
// BodyH is a middleware that takes a struct and a request handler. 
// It parses the request body into the given struct and pushes it into 
// the request's context using the BodyCtxKey`

// BodyF is a middleware that takes a struct and a controller function. 
// It parses the request body into the given struct and pushes it into 
// the request's context using the BodyCtxKey
```

File: middleware/v1/body.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+
+	"github.com/opensaucerer/gotemplate/helper"
+	"github.com/opensaucerer/gotemplate/typing"
+)
+
+//BodyH is a middleware that takes a struct and a request handler. It parses
```
Comment (opensaucerer) [inline middleware/v1/body.go]: Please let the comment be structured the way I structured it in my comment.

Comment (Funmi4194) [reply middleware/v1/body.go]: Done.
I apologize, I didn’t see the structure you sent early, I only got the comment in original text.

File: middleware/v1/body.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+			//append decoded data to request context
+			ctx := context.WithValue(r.Context(), typing.BodyCtxKey{}, bodyStruct)
+
+			//call next handler
+			next.ServeHTTP(w, r.WithContext(ctx))
+		})
+	}
+}
+
+//BodyF is a middleware that takes a struct and a controller function.
```
Comment (opensaucerer) [inline middleware/v1/body.go]: Please just copy and paste...this still doesn't follow

```go
// BodyH is a middleware that takes a struct and a request handler. 
// It parses the request body into the given struct and pushes it into 
// the request's context using the BodyCtxKey`

// BodyF is a middleware that takes a struct and a controller function. 
// It parses the request body into the given struct and pushes it into 
// the request's context using the BodyCtxKey
```

Comment (Funmi4194) [reply middleware/v1/body.go]: done

**Review by opensaucerer — APPROVED** [review]
LGTM

**Pull request merged**